### PR TITLE
gh-111726: Explicitly close database connections in sqlite3 doctests

### PR DESCRIPTION
**Pull request by sobolevn**
My logic behind this was:
- Never add `.close()` to `.. doctest::`, because it distracts people from the core example (except for cases, where it was already present)
- Always add `.close()` to `.. testcode::`, because one code part is better than two

* Issue: gh-111726

----
:books: Documentation preview :books:: https://cpython-previews--111730.org.readthedocs.build/

### CODE REVIEW

**Comment by sobolevn** [comment]
There are still some left:

```
 Document: library/sqlite3
-------------------------
Exception ignored in: <sqlite3.Connection object at 0x7f4d99de1590>
Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Lib/functools.py", line 36, in update_wrapper
    def update_wrapper(wrapper,
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x7f4d99de1590>
Exception ignored in: <sqlite3.Connection object at 0x7f4d99de3ce0>
Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Lib/functools.py", line 36, in update_wrapper
    def update_wrapper(wrapper,
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x7f4d99de3ce0>
Exception ignored in: <sqlite3.Connection object at 0x7f4d9dc24160>
Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Lib/functools.py", line 36, in update_wrapper
    def update_wrapper(wrapper,
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x7f4d9dc24160>
Exception ignored in: <sqlite3.Connection object at 0x7f4d9dc246b0>
Traceback (most recent call last):
  File "/home/runner/work/cpython/cpython/Lib/functools.py", line 36, in update_wrapper
    def update_wrapper(wrapper,
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x7f4d9dc246b0>
4 items passed all tests:
  72 tests in default
   3 tests in sqlite3.cursor
   3 tests in sqlite3.limits
   8 tests in sqlite3.trace
86 tests in 4 items.
86 passed and 0 failed.
Test passed.
1 items passed all tests:
   9 tests in default (cleanup code)
9 tests in 1 items.
9 passed and 0 failed.
Test passed.
Exception ignored in sys.unraisablehook: <function debug at 0x7f4da10aec90>
Traceback (most recent call last):
  File "<doctest sqlite3.trace[4]>", line 2, in debug
AttributeError: 'sqlite3.Connection' object has no attribute '__name__'
Exception ignored in sys.unraisablehook: <function debug at 0x7f4da10aec90>
Traceback (most recent call last):
  File "<doctest sqlite3.trace[4]>", line 2, in debug
AttributeError: 'sqlite3.Connection' object has no attribute '__name__'
```

**Comment by sobolevn** [comment]
One more thing: when running locally I got these files in my `Doc/` folder:

```
» git status
On branch issue-111735
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        new file:   Doc/backup.db
        new file:   Doc/dump.sql
        new file:   Doc/example.db
        modified:   Doc/library/turtle.rst
        new file:   Doc/tutorial.db
```

**Comment by erlend-aasland** [comment]
> One more thing: when running locally I got these files in my `Doc/` folder:
> 
> 
> 
> ```
> 
> » git status
> 
> On branch issue-111735
> 
> Changes to be committed:

Yeah, we don't clean up temporary files in the doctests. We should, though.

**Comment by sobolevn** [comment]
@erlend-aasland do you have any idea why do we still have ResourceWarnings? I've looked through the code multiple time, but I cannot find leaking resources.

**Comment by erlend-aasland** [comment]
> @erlend-aasland do you have any idea why do we still have ResourceWarnings? I've looked through the code multiple time, but I cannot find leaking resources.

I'll have a look tomorrow.

**Comment by sobolevn** [comment]
Ok, looks like I understood the problem.

Here's the sample code that does this (it is in `sqlite3.rst`):

```python
      >>> sqlite3.enable_callback_tracebacks(True)
      >>> con = sqlite3.connect(":memory:")
      >>> def evil_trace(stmt):
      ...     5/0
      ...
      >>> con.set_trace_callback(evil_trace)
      >>> def debug(unraisable):
      ...     print(f"{unraisable.exc_value!r} in callback {unraisable.object.__name__}")
      ...     print(f"Error message: {unraisable.err_msg}")
      >>> import sys
      >>> sys.unraisablehook = debug
      >>> cur = con.execute("SELECT 1")
      ZeroDivisionError('division by zero') in callback evil_trace
      Error message: None
```

**Comment by sobolevn** [comment]
There are still lots of warnings to fix:

```
Document: library/sqlite3
-------------------------
4 items passed all tests:
  71 tests in default
   3 tests in sqlite3.cursor
   3 tests in sqlite3.limits
   7 tests in sqlite3.trace
84 tests in 4 items.
84 passed and 0 failed.
Test passed.
2 items passed all tests:
  12 tests in default (cleanup code)
   1 tests in sqlite3.trace (cleanup code)
13 tests in 2 items.
13 passed and 0 failed.
Test passed.
Exception ignored in: <sqlite3.Connection object at 0x127adbf00>
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython2/Doc/venv/lib/python3.13/site-packages/docutils/nodes.py", line 387, in __new__
    def __new__(cls, data, rawsource=None):
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x127adbf00>
Exception ignored in: <sqlite3.Connection object at 0x12740ce20>
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython2/Doc/venv/lib/python3.13/site-packages/docutils/nodes.py", line 387, in __new__
    def __new__(cls, data, rawsource=None):
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x12740ce20>
Exception ignored in: <sqlite3.Connection object at 0x12740ef10>
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython2/Doc/venv/lib/python3.13/site-packages/docutils/nodes.py", line 387, in __new__
    def __new__(cls, data, rawsource=None):
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x12740ef10>
Exception ignored in: <sqlite3.Connection object at 0x1242f9bf0>
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython2/Doc/venv/lib/python3.13/site-packages/docutils/nodes.py", line 387, in __new__
    def __new__(cls, data, rawsource=None):
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x1242f9bf0>
Exception ignored in: <sqlite3.Connection object at 0x123539590>
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython2/Doc/venv/lib/python3.13/site-packages/docutils/nodes.py", line 387, in __new__
    def __new__(cls, data, rawsource=None):
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x123539590>

Document: library/hashlib
-------------------------
1 items passed all tests:
  58 tests in default
58 tests in 1 items.
58 passed and 0 failed.
Test passed.

Document: library/binascii
--------------------------
1 items passed all tests:
   5 tests in default
5 tests in 1 items.
5 passed and 0 failed.
Test passed.
Exception ignored in: <sqlite3.Connection object at 0x1242faad0>
Traceback (most recent call last):
  File "/Users/sobolev/Desktop/cpython2/Doc/venv/lib/python3.13/site-packages/docutils/nodes.py", line 387, in __new__
    def __new__(cls, data, rawsource=None):
ResourceWarning: unclosed database in <sqlite3.Connection object at 0x1242faad0>
```

**Comment by sobolevn** [comment]
Now I give up :)
No ideas left.

File: Doc/library/sqlite3.rst [hunk 1]
```diff
@@ -1069,6 +1089,7 @@ Connection objects
          Added the ``sqlite3.enable_load_extension`` auditing event.
 
       .. testsetup:: sqlite3.loadext
+         :skipif: True
```
Comment (vstinner) [inline Doc/library/sqlite3.rst]: What's the point of having a test if it's always skipped? Remove it if it always fail?

**Comment by vstinner** [comment]
> @erlend-aasland do you have any idea why do we still have ResourceWarnings?

You can run tests with tracemalloc to see where the object triggering ResourceWarnings was created: https://docs.python.org/dev/library/devmode.html#resourcewarning-example

**Comment by vstinner** [comment]
> Never add .close() to .. doctest::, because it distracts people from the core example (except for cases, where it was already present)

I would prefer to make it explicit that calling close() is recommended. That's why ResourceWarning is emitted if it's not called.

**Comment by erlend-aasland** [comment]
> I would prefer to make it explicit that calling close() is recommended. That's why ResourceWarning is emitted if it's not called.

We should indeed put more emphasis on that in the docs, but there is no reason to do it in _every_ example. There's a lot of examples in the sqlite3 docs, and filling each and everyone of them with best practises is IMO not a good idea for well written docs :)

File: Doc/library/sqlite3.rst [hunk 2]
```diff
@@ -1098,14 +1119,17 @@ Connection objects
...
-         con.close()
-
       .. testoutput:: sqlite3.loadext
          :hide:
 
          (2, 'broccoli pie', 'broccoli cheese onions flour')
          (3, 'pumpkin pie', 'pumpkin sugar flour butter')
 
+      .. testcleanup::
+         :skipif: True
```
Comment (vstinner) [inline Doc/library/sqlite3.rst]: What is the purpose of this testcleanup if it's always skipped?

**Review by vstinner — APPROVED** [review]
LGTM. But I don't get the two `:skipif: True`. Does it mean that the cleanup is never executed? If yes, just remove it, no?

@erlend-aasland: Are you ok with this PR?

**Comment by vstinner** [comment]
@sobolevn @erlend-aasland: Ping

**Comment by erlend-aasland** [comment]
I'll get to this soon!

File: Doc/library/sqlite3.rst [hunk 3]
```diff
@@ -221,6 +223,13 @@ creating a new cursor, then querying the database:
    >>> print(f'The highest scoring Monty Python movie is {title!r}, released in {year}')
    The highest scoring Monty Python movie is 'Monty Python and the Holy Grail', released in 1975
 
+.. testcleanup::
+
+   new_con.close()
+
+   import os
+   os.remove("tutorial.db")
```
Comment (vstinner) [inline Doc/library/sqlite3.rst]: Would it be possible to use support.os_helper.unlink() here?

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I don't know. There is no precedent for it elsewhere in `Doc/`.

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Resolved with #117604 for now.

**Comment by erlend-aasland** [comment]
I tore out some of my commits as a separate PR:

- #117623

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by vstinner** [comment]
I'm not sure what's going on with this PR. Once I saw "os.unlink()" calls, then they disappears. Does @sobolevn or @erlend-aasland plan to propose a following PR to call "os.unlink()"? I proposed using support.os_helper.unlink() which is more reliable when used in tests.

**Comment by erlend-aasland** [comment]
> I'm not sure what's going on with this PR. Once I saw "os.unlink()" calls, then they disappears. Does @sobolevn or @erlend-aasland plan to propose a following PR to call "os.unlink()"? I proposed using support.os_helper.unlink() which is more reliable when used in tests.

They were fixed by Hugo in #117604, as I already said in https://github.com/python/cpython/pull/111730#discussion_r1555295525.

`support.os_helper.unlink` is not used in any doctests at the moment. Perhaps we should rewrite all relevant doctests to use `support.os_helper.unlink`?

**Comment by erlend-aasland** [comment]
@vstinner: Sphinx recommends using its own machinery (`doctest_global_cleanup` and `testcleanup`) for cleaning up temporary doctest files:

> You can use this to e.g. remove any temporary files that the tests leave behind.

**Comment by erlend-aasland** [comment]
We cannot land this yet, there are still warnings to be fixed, as Nikita said in https://github.com/python/cpython/pull/111730#issuecomment-1798046622.

**Comment by vstinner** [comment]
> They were fixed by Hugo in https://github.com/python/cpython/pull/117604,

Great!

> as I already said in https://github.com/python/cpython/pull/111730#discussion_r1555295525.

Oh, I'm lost in GitHub UI which hides comments when a comment is "solved".

> support.os_helper.unlink is not used in any doctests at the moment. Perhaps we should rewrite all relevant doctests to use support.os_helper.unlink?

Nah, it's not worth it, os.unlink() is fine. It was just a comment on a review.

> We cannot land this yet, there are still warnings to be fixed, as Nikita said in https://github.com/python/cpython/pull/111730#issuecomment-1798046622.

I suggest to merge these fixes to make the situation clearer, since I'm lost in the lost history of this PR. You can, just remove "all" from the PR title :-)

But it's up to you, if you want to first really fix all warnings.

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by erlend-aasland** [comment]
@sobolevn, @vstinner: I fixed the remaining issues with f170f7ccb53cb6ac9e115bee306ed37b20630088. It seems Sphinx has a problem with `doctest` and following `testcleanup`. I chose to fix the issues by adding more explicit `.close()`s to `doctest`s. It is not bad to promote good behaviour more explicitly.

cc. @AA-Turner

**Comment by erlend-aasland** [comment]
Sorry it took so long, Nikita!

**Comment by sobolevn** [comment]
@erlend-aasland thanks a lot for your help and taking this over! 🤝

**Pull request merged**